<h1>Pre-procesamiento y limpieza de Datos</h1>

In [1]:
'''Importar la libreria para la realizacion de este laboratorio.'''
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

In [2]:
'''Obtener el Dataset para mostrar todo su contenido.'''
df = pd.read_csv("data_act_01.csv",sep=";")
df.sample(10)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
3930  160932871                   909  2016-04-02T00:00:00    18:14   
9422  160961548          Passing Call  2016-04-05T00:00:00    11:54   
2573  160930454            Trespasser  2016-04-02T00:00:00     3:23   
9651  160962539    Homeless Complaint  2016-04-05T00:00:00    15:47   
2282  160924023    Homeless Complaint  2016-04-01T00:00:00    23:45   
1015  160921688          Passing Call  2016-04-01T00:00:00    13:00   
3799  160932629         Audible Alarm  2016-04-02T00:00:00    17:13   
4602  160940049        Noise Nuisance  2016-04-03T00:00:00     0:15   
2952  160931135                22500e  2016-04-02T00:00:00    10:06   
1379  160922434     Suspicious Person  2016-04-01T00:00:00    16:18   

             CallDateTime   Disposition                   Address  \
3930  2016-04-02T18:14:00           NOM    3600 Block Of Geary Bl   
9422  2016-04-05T11:54:00           HAN       0 Block Of Peace Pz   
2573  2016-04-02T03:23:00           ADV   300 Block Of Battery St   
9651  2016-04-05T15:47:00           ADV  1300 Block Of Mission St   
2282  2016-04-01T23:45:00           ADV    700 Block Of Natoma St   
1015  2016-04-01T13:00:00  Not recorded         3rd St/quesada Av   
3799  2016-04-02T17:13:00           CAN   300 Block Of Shipley St   
4602  2016-04-03T00:15:00           CAN     500 Block Of Minna St   
2952  2016-04-02T10:06:00           GOA   0 Block Of Edinburgh St   
1379  2016-04-01T16:18:00           HAN      100 Block Of Eddy St   

               City State AgencyId  Range      AddressType  
3930  San Francisco    CA        1    NaN  Premise Address  
9422  San Francisco    CA        1    NaN  Common Location  
2573  San Francisco    CA        1    NaN  Premise Address  
9651  San Francisco    CA        1    NaN  Premise Address  
2282  San Francisco    CA        1    NaN  Premise Address  
1015  San Francisco    CA        1    NaN     Intersection  
3799  San Francisco    CA        1    NaN  Premise Address  
4602  San Francisco    CA        1    NaN  Premise Address  
2952  San Francisco    CA        1    NaN  Premise Address  
1379  San Francisco    CA        1    NaN  Premise Address

In [3]:
'''Observar el tipo de variables que contiene el dataframe'''
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CrimeId                10051 non-null  int64  
 1   OriginalCrimeTypeName  10051 non-null  object 
 2   OffenseDate            10051 non-null  object 
 3   CallTime               10051 non-null  object 
 4   CallDateTime           10051 non-null  object 
 5   Disposition            10051 non-null  object 
 6   Address                10051 non-null  object 
 7   City                   9730 non-null   object 
 8   State                  10048 non-null  object 
 9   AgencyId               10051 non-null  object 
 10  Range                  0 non-null      float64
 11  AddressType            10051 non-null  object 
dtypes: float64(1), int64(1), object(10)
memory usage: 942.4+ KB


In [4]:
'''Detectar valores nulos y su conteo'''
df.isnull().sum()

CrimeId                      0
OriginalCrimeTypeName        0
OffenseDate                  0
CallTime                     0
CallDateTime                 0
Disposition                  0
Address                      0
City                       321
State                        3
AgencyId                     0
Range                    10051
AddressType                  0
dtype: int64

In [5]:
'''Observar la cantidad de filas en el dataset.'''
df.shape

(10051, 12)

# Primer error (Columna Range)

In [6]:
'''Eliminar la columna "Range" ya que observando los registros se observa que no posee ningun dato.'''
df = df.drop(['Range'], axis=1)
df.sample(5)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
2548  160930402          Traffic Stop  2016-04-02T00:00:00     2:55   
4029  160933056                22500f  2016-04-02T00:00:00    19:11   
6742  160943790          Traffic Stop  2016-04-03T00:00:00    23:45   
8546  160953255     Suspicious Person  2016-04-04T00:00:00    18:43   
1156  160922013                22500e  2016-04-01T00:00:00    14:26   

             CallDateTime Disposition                  Address           City  \
2548  2016-04-02T02:55:00         CIT    3300 Block Of 21st St  San Francisco   
4029  2016-04-02T19:11:00         GOA  300 Block Of Madison St  San Francisco   
6742  2016-04-03T23:45:00         CIT       Fell St/shrader St  San Francisco   
8546  2016-04-04T18:43:00         ADV   500 Block Of Bryant St  San Francisco   
1156  2016-04-01T14:26:00         CIT  1100 Block Of Moraga St  San Francisco   

     State AgencyId      AddressType  
2548    CA        1  Premise Address  
4029    CA        1  Premise Address  
6742    CA        1     Intersection  
8546    CA        1  Premise Address  
1156    CA        1  Premise Address

# Segundo error (IDs duplicados)

In [7]:
'''Conteo de ids duplicados, y verificacion de los mismos.'''
print("CrimeId: ",df['CrimeId'].duplicated().sum())
dup = df[df['CrimeId'].duplicated(keep=False)]['CrimeId'].tolist()
dup

CrimeId:  4


[160913455, 160913455, 160913455, 160950496, 160950496, 160950496]

# Tercer error (Rellenar ciudades por valores)

In [8]:
'''Se reemplaza los valores vacios que existen 
en las columnas City por valores en Cero, 
ya que al tener valores nulos pueden generar
errores adicionales en este contexto.'''

df[['City', 'State']] = df[['City','State']].fillna(value=0)
df

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

              CallDateTime   Disposition                  Address  \
0      2016-03-30T18:42:00           REP  100 Block Of Chilton Av   
1      2016-03-31T15:31:00           GOA  2300 Block Of Market St   
2      2016-03-31T16:49:00           GOA  2300 Block Of Market St   
3      2016-03-31T17:38:00           GOA      500 Block Of 7th St   
4      2016-03-31T17:42:00           REP       Beale St/bryant St   
...                    ...           ...                      ...   
10046  2016-04-05T23:38:00           GOA    Hollister Av/hawes St   
10047  2016-04-05T23:43:00           ADV   California St/mason St   
10048  2016-04-05T23:48:00           CIT   Rosella Ct/onondaga Av   
10049  2016-04-05T23:48:00  Not recorded          0 Blk Herbst Rd   
10050  2016-04-05T23:54:00           ADV        Grant Av/geary St   

                City State AgencyId      AddressType  
0      San Francisco    CA        1  Premise Address  
1      San Francisco    CA        1  Premise Address  
2      San Francisco    CA        1  Premise Address  
3      San Francisco    CA        1  Premise Address  
4      San Francisco    CA        1     Intersection  
...              ...   ...      ...              ...  
10046  San Francisco    CA        1     Intersection  
10047  San Francisco    CA        1     Intersection  
10048  San Francisco    CA        1     Intersection  
10049  San Francisco    CA        1     Geo-Override  
10050  San Francisco    CA        1     Intersection  

[10051 rows x 11 columns]

In [9]:
'''Verificar la redundancia de datos en la columna de City'''
df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
0                  321                    321          321       321   
 S                   1                      1            1         1   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
SAN FRANCISCO        1                      1            1         1   
San Francisco     9665                   9665         9665      9665   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  \
City                                                                 
0                       321          321      321    321       321   
 S                        1            1        1      1         1   
Brisbane                  1            1        1      1         1   
Daly City                 5            5        5      5         5   
Presidio                  3            3        3      3         3   
SAN FRANCISCO             1            1        1      1         1   
San Francisco          9665         9665     9665   9665      9665   
Treasure Isla            51           51       51     51        51   
Yerba Buena               3            3        3      3         3   

               AddressType  
City                        
0                      321  
 S                       1  
Brisbane                 1  
Daly City                5  
Presidio                 3  
SAN FRANCISCO            1  
San Francisco         9665  
Treasure Isla           51  
Yerba Buena              3

<h1>Analisis</h1>
Observamos que existian 321 valores nulos que fueron reemplazados por 0 Analizando las direccciones de estas filas, nos podemos dar cuenta que todas pertenecen a la ciudad de San Francisco, asi que reemplazaremos estos Valores por San Francisco. Del mismo modo se observan dos errores, existe una columna con la letra S, probablemente de San Francisco, comprobaremos la direccion de esta para estar seguros de que la ciudad es San Francisco. Asi mismo observamos que existe un dato con mayusculas. Vamos a corregir ese error.

In [10]:
'''Se reemplaza el dato en mayusculas, por minusculas.'''
df['City'] = df['City'].replace('SAN FRANCISCO','San Francisco');

In [11]:
'''Se reemplaza ahora la letra S por la ciudad de San Francisco'''
df['City']= df['City'].replace('S','San Francisco')

In [12]:
'''Se reemplaza ahora las filas de 0 por la ciudad de San Francisco'''
df['City']= df['City'].replace(0,'San Francisco')

In [13]:
'''Verificar la redundancia de datos en la columna de City otra vez'''
df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
 S                   1                      1            1         1   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
San Francisco     9987                   9987         9987      9987   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  \
City                                                                 
 S                        1            1        1      1         1   
Brisbane                  1            1        1      1         1   
Daly City                 5            5        5      5         5   
Presidio                  3            3        3      3         3   
San Francisco          9987         9987     9987   9987      9987   
Treasure Isla            51           51       51     51        51   
Yerba Buena               3            3        3      3         3   

               AddressType  
City                        
 S                       1  
Brisbane                 1  
Daly City                5  
Presidio                 3  
San Francisco         9987  
Treasure Isla           51  
Yerba Buena              3

In [14]:
'''Verificamos nuevamente los valores nulos, 
como se observa existen 3 estados que son nulos.'''
df.isnull().sum()

CrimeId                  0
OriginalCrimeTypeName    0
OffenseDate              0
CallTime                 0
CallDateTime             0
Disposition              0
Address                  0
City                     0
State                    0
AgencyId                 0
AddressType              0
dtype: int64

# Cuarto error (Valores de columna State)

In [15]:
'''Se verifica la columna state para analizar los valores nulos.
Como se observa existen 3 valores que no tienen valor de estado,
asi que corregiremos este error reemplazando este valor por el estado
correspondiente que es CA.'''
df.groupby('State').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  CallDateTime  \
State                                                                        
0            3                      3            3         3             3   
CA       10048                  10048        10048     10048         10048   

       Disposition  Address   City  AgencyId  AddressType  
State                                                      
0                3        3      3         3            3  
CA           10048    10048  10048     10048        10048

In [18]:
'''Reemplazo de los valores 0 de la columna estado.'''
df['State']= df['State'].replace(0,'CA')

In [19]:
'''Verificar nuevamente'''
df.groupby('State').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  CallDateTime  \
State                                                                        
CA       10051                  10051        10051     10051         10051   

       Disposition  Address   City  AgencyId  AddressType  
State                                                      
CA           10051    10051  10051     10051        10051